service/controller/release/resource/configs/resource.go

```diff
@@ -0,0 +1,174 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/release-operator/v2/pkg/project"
+	"github.com/giantswarm/release-operator/v2/service/controller/key"
+)
+
+const (
+	Name = "configs"
+)
+
+type Config struct {
+	K8sClient k8sclient.Interface
+	Logger    micrologger.Logger
+}
+
+type Resource struct {
+	k8sClient k8sclient.Interface
+	logger    micrologger.Logger
+}
+
+func New(config Config) (*Resource, error) {
+	if config.K8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.K8sClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		k8sClient: config.K8sClient,
+		logger:    config.Logger,
+	}
+
+	return r, nil
+}
+
+func (r *Resource) Name() string {
+	return Name
+}
+
+func (r *Resource) ensureState(ctx context.Context) error {
+	var releases releasev1alpha1.ReleaseList
+	{
+		err := r.k8sClient.CtrlClient().List(
+			ctx,
+			&releases,
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		releases = excludeDeletedRelease(releases)
+		releases = r.excludeUnusedDeprecatedReleases(releases)
+	}
+
+	var components map[string]releasev1alpha1.ReleaseSpecComponent
+	{
+		components = key.ExtractComponents(releases)
+	}
+
+	var configs corev1alpha1.ConfigList
+	{
+		err := r.k8sClient.CtrlClient().List(
+			ctx,
+			&configs,
+			&client.ListOptions{
+				LabelSelector: labels.SelectorFromSet(labels.Set{
+					key.LabelManagedBy: project.Name(),
+				}),
+			},
+		)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	configsToDelete := calculateObsoleteConfigs(components, configs)
+	for _, config := range configsToDelete.Items {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleting config %#q in namespace %#q", config.Name, config.Namespace))
+
+		err := r.k8sClient.CtrlClient().Delete(
+			ctx,
+			&config,
+		)
+		if apierrors.IsNotFound(err) {
+			// fall through.
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("deleted config %#q in namespace %#q", config.Name, config.Namespace))
+	}
+
+	configsToCreate := calculateMissingConfigs(components, configs)
+	for _, config := range configsToCreate.Items {
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating config %#q in namespace %#q", config.Name, config.Namespace))
+
+		err := r.k8sClient.CtrlClient().Create(
+			ctx,
+			&config,
+		)
+		if apierrors.IsAlreadyExists(err) {
+			// fall through.
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created config %#q in namespace %#q", config.Name, config.Namespace))
+	}
+
+	return nil
+}
+
+func (r *Resource) excludeUnusedDeprecatedReleases(releases releasev1alpha1.ReleaseList) releasev1alpha1.ReleaseList {
```

Comment:
Can we move these functions to `controller/key/key.go` and then use the same functions in both resources? IIRC these are already in the other resource, right?